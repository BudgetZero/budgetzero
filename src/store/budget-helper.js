import { defineStore } from 'pinia'
import { useBudgetManagerStore } from './pinia'
import { v4 as uuidv4 } from 'uuid'
import validator from 'validator'
import { sanitizeValueInput } from '../helper.js'

export const useBudgetHelperStore = defineStore('budgetHelper', {
  actions: {
    /**
     * Create payee doc.
     * This should only be called from getPayeeID() action.
     * @param {String} payload Plaintext payee name
     * @returns
     */
    createPayee(payload) {
      const budgetManagerStore = useBudgetManagerStore()
      var payee = {
        _id: `b_${budgetManagerStore.budgetID}_payee_${uuidv4()}`,
        name: payload
      }
      return budgetManagerStore.putDocument(payee)
    },

    /**
     * Returns the payee UUID for any payee name. Dispatches action to create the payee if it doesn't exist yet.
     * @param {String} payload Plaintext payee name. e.g. 'Grocery Store'
     * @returns Payee UUID
     */
    async getPayeeID(payload) {
      const budgetManagerStore = useBudgetManagerStore()

      //First, check if this payee has already been created
      const payeeLookup = Object.keys(budgetManagerStore.payee_map).find(
        (key) => budgetManagerStore.payee_map[key] === payload
      )

      if (payeeLookup) {
        return payeeLookup
      } else if (validator.isUUID(`${payload}`)) {
        // If the payload is already UUID then return.
        return payload
      } else if (payload === '---------------------initial-balance') {
        //If it's initial balance then return
        return payload
      } else if (typeof payload === 'undefined' || payload === null || payload === '') {
        // If payload is an object, then it's an existing payee. Otherwise we need to create the payee.
        return null
      } else if (typeof payload != 'string') {
        return payload._id
      } else {
        // Payload is a string. Need to create payee to get an uuid
        let payee = await this.createPayee(payload)
        return payee.id.slice(-36)
      }
    },

    /**
     * Create/update the mirrored transfer transaction
     */
    saveMirroredTransferTransaction(context, payload) {
      var mirroredTransferTransaction = Object.assign({}, payload)

      //Check if the mirrored transaction doesn't exist then we create
      if (payload.transfer) {
        const index =
          context.getters.transactionsLookupByID[
            `b_${context.getters.selectedBudgetID}_transaction_${payload.transfer}`
          ]

        mirroredTransferTransaction = Object.assign({}, context.getters.transactions[index])
      } else {
        //Creating new transaction
        mirroredTransferTransaction._id = `b_${
          context.getters.selectedBudgetID
        }_transaction_${Vue.prototype.$vm.$uuid.v4()}`

        delete mirroredTransferTransaction._rev
      }
      //Create the mirrored transaction
      mirroredTransferTransaction.value = -payload.value
      mirroredTransferTransaction.transfer = payload._id.slice(-36)
      mirroredTransferTransaction.account = payload.payee
      mirroredTransferTransaction.payee = payload.account //The payee is the _id of the other account
      mirroredTransferTransaction.memo = payload.memo
      mirroredTransferTransaction.category = null
      mirroredTransferTransaction.date = payload.date
      mirroredTransferTransaction.cleared = payload.cleared

      context.dispatch('commitDocToPouchAndVuex', mirroredTransferTransaction)

      return mirroredTransferTransaction._id.slice(-36)
    },

    /**
     * Create or update transaction
     * @param {doc} payload The transaction to create or update
     */
    async putTransaction(payload) {
      const budgetManagerStore = useBudgetManagerStore()

      //Check if this is a transfer transaction. if so, get the account ID
      //TODO: only let this be a transfer if the account actually exists?
      if (payload.payee && payload.payee.includes('Transfer: ')) {
        //account_id is the account the original transfer is going to
        const account_id = Object.keys(budgetManagerStore.account_map).find(
          (key) => budgetManagerStore.account_map[key] === payload.payee.slice(10)
        )

        payload.payee = account_id
        const mirroredTransferID = await context.dispatch('saveMirroredTransferTransaction', payload)
        payload.transfer = mirroredTransferID
        payload.category = null
      } else {
        payload.transfer = null
      }

      payload.value = sanitizeValueInput(payload.value)

      await this.getPayeeID(payload.payee).then((response) => {
        payload.payee = response
      })
      return budgetManagerStore.putDocument(payload)
    }
  }
})
