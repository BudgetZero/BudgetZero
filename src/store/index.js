import Vue from "vue";
import Vuex from "vuex";
import user from "./modules/user-module";
import budget from "./modules/budget-module"
import pouchdb from "./modules/pouchdb-module";
import moment from "moment";

Vue.use(Vuex);

export default new Vuex.Store({
  strict: process.env.NODE_ENV !== 'production',
  modules: {
    user,
    budget,
    pouchdb,
  },
  state: {
    error_msg: "",
    snackbar: false,
    sync_state: "",
    selectedBudgetID: null,
    month_selected: moment(new Date()).format("YYYY-MM"),
  },
  getters: {
    error_msg: state => state.error_msg,
    sync_state: state => state.sync_state,
    snackbar: state => state.snackbar,
    selectedBudgetID: state => state.selectedBudgetID,
    month_selected: state => state.month_selected,
  },
  mutations: {
    SET_STATUS_MESSAGE(state, message) {
      // console.log('failure');
      state.sync_state = `${message}`;
    },
    SET_ERROR_MESSAGE(state, error) {
      state.error_msg = error
      state.snackbar = true
    },
    SET_SNACKBAR(state, snackbar) {
      state.snackbar = snackbar
    },
    ADD_MONTH(state) {
      state.month_selected = moment(state.month_selected)
        .add(1, "M")
        .format("YYYY-MM");
    },
    PREVIOUS_MONTH(state) {
      state.month_selected = moment(state.month_selected)
        .subtract(1, "M")
        .format("YYYY-MM");
    },
    GO_TO_CURRENT_MONTH(state) {
      state.month_selected = moment(new Date()).format("YYYY-MM");
    },
    UPDATE_SELECTED_BUDGET(state, payload) {
      state.selectedBudgetID = payload;
      localStorage.budgetID = payload;
    },
  },
  actions: {
    setSnackBarBoolean(context, snackbar) {
      context.commit('SET_SNACKBAR', snackbar)
    },
  }
});
