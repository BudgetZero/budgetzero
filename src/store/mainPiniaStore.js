import { defineStore } from 'pinia'
import moment from 'moment'

export const useMainStore = defineStore('mainPinia', {
  state: () => {
    return {
      snackBarMessage: '',
      snackBarColor: '',
      snackBar: false,
      sync_state: '',
      selectedBudgetID: null,
      month_selected: moment(new Date()).format('YYYY-MM'),
      manageBudgetOverlay: true,
      loadingOverlay: true,
    }
  },
  actions: {
    setSnackbarMessage(payload) {
      this.snackBarMessage = payload.snackBarMessage
      this.snackBarColor = payload.snackBarColor

      this.snackBar = true
    },
    addMonth() {
      this.month_selected = moment(this.month_selected).add(1, 'M').format('YYYY-MM')
    },
    subtractMonth() {
      this.month_selected = moment(this.month_selected).subtract(1, 'M').format('YYYY-MM')
    },
    setCurrentMonth() {
      this.month_selected = moment(new Date()).format('YYYY-MM')
    },
  }
})