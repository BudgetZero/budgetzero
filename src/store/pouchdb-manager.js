// import { resolve } from 'core-js/fn/promise'
import _, { reject } from 'lodash'
import moment from 'moment'
import PouchDB from 'pouchdb'
import {
  schema_budget,
  schema_budget_opened,
  schema_account,
  schema_transaction,
  schema_category,
  schema_m_category,
  schema_masterCategory,
  schema_payee,
  validateSchema
} from './validation'

export class PouchDBManager {
  constructor() {
    this.localdb = new PouchDB(new Date().toISOString())
    this.remoteSyncDB = null
    this.syncHandler = null
  }

  setRemoteSyncDB(remoteURL) {
    this.remoteSyncDB = new PouchDB(remoteURL)
  }

  startRemoteSync() {
    if (this.remoteSyncDB) {
      this.syncHandler = PouchDB.sync(this.remoteSyncDB, this.localdb, { live: true })
        .on('paused', function (info) {
          console.info('replication paused', info)
        })
        .on('active', function () {
          console.info('replication active')
        })
        .on('denied', function (err) {
          console.error('replication denied', err)
        })
        .on('error', function (err) {
          console.error('replication error', err)
        })
    } else {
      console.error('No remote sync URL found')
    }
  }

  stopRemoteSync() {
    this.syncHandler.cancel()
  }
}
