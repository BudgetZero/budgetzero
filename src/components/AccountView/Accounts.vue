<template>
  <v-card elevation="0" class="mx-auto">
    <v-data-table
      id="accountsTable"
      :headers="headers"
      :items="budgetManagerStore.accounts"
      sort-by="calories"
      class="elevation-1 account-table"
      hide-default-footer
      disable-pagination
    >
      <template #top>
        <v-toolbar flat color="white">
          <span class="text-h3">Accounts</span>
          <v-spacer />

          <AccountAddModal v-model="showModal" :editeditem="editedItem" @save="save()" />

          <v-btn id="addAccountBtn" color="accent" dark class="mb-2" @click="create()"> Add Account </v-btn>
        </v-toolbar>
        <v-divider class="pb-4" />
      </template>

      <template #item.action="{ item }">
        <div class="crud-actions">
          <v-icon icon dark class="" color="primary" @click="editItem(item)"> edit </v-icon>
          <v-icon icon dark class="ml-1" color="accent" @click="deleteItem(item)"> delete </v-icon>
        </div>
      </template>
      <template #no-data>
        <!-- <v-btn color="primary" @click="on">Add Account</v-btn> -->
      </template>
    </v-data-table>
  </v-card>
</template>
<script>
import AccountAddModal from './AccountAddModal.vue'
import { mapStores } from 'pinia'
import { useBudgetManagerStore } from '../../store/budgetManager'
import { useBudgetHelperStore } from '../../store/budgetManagerHelper'
import { useMainStore } from '../../store/mainPiniaStore'

export default {
  name: 'AccountGrid',
  components: {
    AccountAddModal
  },
  data() {
    return {
      headers: [
        {
          text: 'Name',
          align: 'left',
          sortable: false,
          value: 'name'
        },
        { text: 'Type', value: 'type' },
        { text: 'On Budget', value: 'onBudget' },
        { text: 'Invert Balance', value: 'balanceIsNegative' },
        { text: 'Closed', value: 'closed' },
        { text: 'Actions', value: 'action', sortable: false }
      ],
      editedIndex: -1,
      editedItem: null,
      emptyItem: {
        type: '',
        checkNumber: true,
        closed: false,
        name: '',
        note: null,
        sort: 0,
        onBudget: true,
        balanceIsNegative: false,
        initialBalance: 0
      },
      showModal: false,
      nameRules: [
        (v) => !!v || 'Name is required',
        (v) => (v && v.length <= 10) || 'Name must be less than 10 characters'
      ]
    }
  },
  computed: {
    ...mapStores(useBudgetManagerStore, useBudgetHelperStore, useMainStore)
  },
  mounted() {},
  created() {},
  methods: {
    create() {
      this.editedIndex = -1
      this.editedItem = Object.assign({}, this.emptyItem)
      this.showModal = true
    },
    editItem(item) {
      this.editedIndex = this.budgetManagerStore.accounts.indexOf(item)
      this.editedItem = Object.assign({}, item)
      this.showModal = true
    },
    async deleteItem(item) {
      const index = this.budgetManagerStore.accounts.indexOf(item)

      await this.$root.$confirm('Delete Account', 'Are you sure you want to delete this account?', {
        onlyShowAgreeBtn: false,
        agreeBtnColor: 'accent',
        agreeBtnText: 'Delete Account',
        cancelBtnColor: 'grey'
      })

      try {
        await this.budgetHelperStore.deleteAccount(this.budgetManagerStore.accounts[index])
      } catch (err) {
        await this.$root.$confirm(
          'Deletion Failed',
          `This account still has ${err.message} transaction(s). You must delete those transactions to delete the account.`,
          {
            onlyShowAgreeBtn: true,
            agreeBtnColor: 'accent',
            agreeBtnText: 'Ok'
          }
        )
      }
    },
    close() {
      this.showModal = false
      setTimeout(() => {
        // this.editedItem = Object.assign({}, this.defaultItem);
        this.editedIndex = -1
      }, 300)
    },
    async save() {
      if (this.editedIndex > -1) {
        // Editing existing account
        const editPayload = {
          type: this.editedItem.type,
          checkNumber: this.editedItem.checkNumber,
          closed: this.editedItem.closed,
          name: this.editedItem.name,
          note: this.editedItem.note,
          sort: this.editedItem.sort,
          onBudget: this.editedItem.onBudget,
          balanceIsNegative: this.editedItem.balanceIsNegative,
          _id: this.editedItem._id,
          _rev: this.editedItem._rev
        }
        Object.assign(this.budgetManagerStore.accounts[this.editedIndex], this.editedItem)
        await this.budgetHelperStore.createUpdateAccount(editPayload, false)
      } else {
        // Creating new account
        const newPayload = {
          type: this.editedItem.type,
          checkNumber: this.editedItem.checkNumber,
          closed: this.editedItem.closed,
          name: this.editedItem.name,
          note: this.editedItem.note,
          sort: this.editedItem.sort,
          onBudget: this.editedItem.onBudget,
          balanceIsNegative: this.editedItem.balanceIsNegative,
          _id: `b_${this.budgetManagerStore.budgetID}_account_${this.$uuid.v4()}`
        }
        await this.budgetHelperStore.createUpdateAccount(newPayload, this.editedItem.initialBalance)
      }
      this.close()
    }
  }
}
</script>

<style scoped></style>
