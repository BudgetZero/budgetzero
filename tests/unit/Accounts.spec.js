import { shallowMount, mount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Accounts from "@/components/AccountView/Accounts.vue";
import Vuetify from "vuetify";
import mock_budget from "@/../tests/__mockdata__/mock_budget.json";
import store from "@/store";


import Vue from "vue";
Vue.use(Vuetify);
Vue.config.productionTip = false;

const localVue = createLocalVue();
localVue.use(Vuex);

const $route = {
  path: "/fake/path",
  params: {
    account_id: "38e690f8-198f-4735-96fb-3a2ab15081c2"
  }
};

const data = mock_budget.rows.map(row => row.doc);
let numberOfAccounts = null;

describe("transaction table", () => {
  let vuetify;
  let wrapper;

  beforeAll(() => {
    // state.monthCategoryBudgets = data.filter(row => row._id.includes("_m_category_"));
    // state.payees = data.filter(row => row._id.includes("_payee_"));

    store.state.pouchdb.accounts = data.filter(row => row._id.includes("_account_"));
    store.state.pouchdb.transactions = data.filter(row => row._id.includes("_transaction_"));
    store.state.pouchdb.masterCategories = data.filter(row =>
      row._id.includes("_master-category_")
    );
    store.state.pouchdb.categories = data
      .filter(row => row._id.includes("_category_"))
      .filter(row => !row._id.includes("m_category"));
    store.state.pouchdb.month_selected = "2020-12",
    store.state.pouchdb.selectedBudgetID = "79de488f-448e-4b4d-97ad-61e5e4f5df31",

    numberOfAccounts = data.filter(row => row._id.includes("_account_")).length; 
  });

  beforeEach(() => {
    vuetify = new Vuetify();
    localVue.use(vuetify);

    store.state.month_selected = "2020-12";
    wrapper = mount(Accounts, {
      store,
      localVue,
      vuetify,
      mocks: {
        $route
      },
      stubs: {
        ImportModalComponent: true
      }
    });
  });

  it("account renders correct snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("account table exists", () => {
    expect(wrapper.find("#accountsTable").exists()).toBeTruthy();
  });

  //   it("accounts vuex getter is correct length", () => {
  //     expect(wrapper.vm.accounts.length).toBe(numberOfAccounts)
  //   });

  it("clicking Add Account button shows modal", async () => {
    // Open modal
    wrapper.find("#addAccountBtn").trigger("click");
    await localVue.nextTick();

    // Verify modal is open
    expect(wrapper.find(".v-dialog .title").text()).toEqual("Edit Account");

  });

  it("submitting add account form", async () => {
    // Open modal
    wrapper.find("#addAccountBtn").trigger("click");
    await localVue.nextTick();

    wrapper.vm.editedItem = {
      type: "CHECKING",
      checkNumber: true,
      closed: false,
      name: "test",
      note: "123",
      sort: 0,
      onBudget: true,
      balanceIsNegative: false,
      initialBalance: 0
    }; 
    wrapper.vm.editedIndex = -1;

    // Trigger save
    wrapper.vm.$store.dispatch = jest.fn();
    
    // wrapper.vm.$uuid.v4() = jest.mockReturnValue('test123123123');
    spyOn(wrapper.vm, "v4").mockReturnValue('123123123')

    // jest.spyOn(wrapper.vm.$uuid, 'v4').mockReturnValueOnce('fake uuid');
    wrapper.find("#saveAccountBtn").trigger("click"); 
    wrapper.vm.save()

    expect(wrapper.vm.$store.dispatch).not.toBeCalled();

  });

  //   it("transaction list for table is correct length", () => {
  //     expect(wrapper.find("#accountsTable > table").length).toBe(7)
  //   });
  // it("add transaction btn creates a new row when clicked", () => {
  //   expect(wrapper.vm.transactionListForTable.length).toBe(numberOfTransactions);
  //   wrapper.find("#addTransactionBtn").trigger('click')
  //   expect(wrapper.vm.transactionListForTable.length).toBe(numberOfTransactions+1);
  // });
});
