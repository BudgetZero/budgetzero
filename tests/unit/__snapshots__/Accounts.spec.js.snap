// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transaction table account renders correct snapshot 1`] = `
<div class="mx-auto v-card v-sheet theme--light elevation-0">
  <div class="v-data-table elevation-1 account-table v-data-table--has-top theme--light" id="accountsTable">
    <header class="v-sheet theme--light v-toolbar v-toolbar--flat white" style="height: 64px;">
      <div class="v-toolbar__content" style="height: 64px;"><span class="text-h3">Accounts</span>
        <div class="spacer"></div>
        <div role="dialog" class="v-dialog__container">
          <!---->
        </div> <button type="button" class="mb-2 v-btn v-btn--contained v-btn--is-elevated v-btn--has-bg theme--dark v-size--default accent" id="addAccountBtn"><span class="v-btn__content">
          Add Account
        </span></button>
      </div>
    </header>
    <hr role="separator" aria-orientation="horizontal" class="pb-4 v-divider theme--light">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Name" class="text-left"><span>Name</span></th>
            <th role="columnheader" scope="col" aria-label="Type: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Type</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="On Budget: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>On Budget</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Invert Balance: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Invert Balance</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Closed: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Closed</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Actions" class="text-start"><span>Actions</span></th>
          </tr>
        </thead>
        <tbody>
          <tr class="">
            <td class="text-left">Savings</td>
            <td class="text-start">CHECKING</td>
            <td class="text-start">true</td>
            <td class="text-start">false</td>
            <td class="text-start">false</td>
            <td class="text-start">
              <div class="crud-actions"><button type="button" icon="" class="v-icon notranslate v-icon--link material-icons theme--dark primary--text">edit</button> <button type="button" icon="" class="v-icon notranslate ml-1 v-icon--link material-icons theme--dark accent--text">delete</button></div>
            </td>
          </tr>
          <tr class="">
            <td class="text-left">Chase</td>
            <td class="text-start">CREDIT</td>
            <td class="text-start">true</td>
            <td class="text-start">true</td>
            <td class="text-start">false</td>
            <td class="text-start">
              <div class="crud-actions"><button type="button" icon="" class="v-icon notranslate v-icon--link material-icons theme--dark primary--text">edit</button> <button type="button" icon="" class="v-icon notranslate ml-1 v-icon--link material-icons theme--dark accent--text">delete</button></div>
            </td>
          </tr>
          <tr class="">
            <td class="text-left">Vanguard</td>
            <td class="text-start">INVESTMENT</td>
            <td class="text-start">false</td>
            <td class="text-start">false</td>
            <td class="text-start">false</td>
            <td class="text-start">
              <div class="crud-actions"><button type="button" icon="" class="v-icon notranslate v-icon--link material-icons theme--dark primary--text">edit</button> <button type="button" icon="" class="v-icon notranslate ml-1 v-icon--link material-icons theme--dark accent--text">delete</button></div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`transaction table submitting add account form 1`] = `
<div class="mx-auto v-card v-sheet theme--light elevation-0">
  <div class="v-data-table elevation-1 account-table v-data-table--has-top theme--light" id="accountsTable">
    <header class="v-sheet theme--light v-toolbar v-toolbar--flat white" style="height: 64px;">
      <div class="v-toolbar__content" style="height: 64px;"><span class="text-h3">Accounts</span>
        <div class="spacer"></div>
        <div role="dialog" class="v-dialog__container">
          <div role="document" tabindex="0" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
            <transition-stub name="dialog-transition" appear="true">
              <div class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 500px;">
                <form novalidate="novalidate" class="v-form">
                  <div class="v-card v-sheet theme--light">
                    <div class="v-card__title primary white--text pb-4"><span class="title"><span>Edit Account</span></span></div>
                    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
                    <div class="v-card__text pb-0">
                      <form novalidate="novalidate" class="v-form">
                        <div class="row">
                          <div class="col col-12">
                            <div class="v-input theme--light v-text-field">
                              <div class="v-input__control">
                                <div class="v-input__slot">
                                  <div class="v-text-field__slot"><label for="input-131" class="v-label theme--light" style="left: 0px; position: absolute;">Account Name</label><input data-cy="account-name" id="input-131" type="text"></div>
                                </div>
                                <div class="v-text-field__details">
                                  <div class="v-messages theme--light">
                                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div class="col col-4">
                            <div class="v-input theme--light v-text-field v-select">
                              <div class="v-input__control">
                                <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-134" class="v-input__slot">
                                  <div class="v-select__slot"><label for="input-134" class="v-label theme--light" style="left: 0px; position: absolute;">Type</label>
                                    <div class="v-select__selections"><input data-cy="account-type" id="input-134" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
                                    <div class="v-input__append-inner">
                                      <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                                    </div><input type="hidden">
                                  </div>
                                  <div class="v-menu">
                                    <!---->
                                  </div>
                                </div>
                                <div class="v-text-field__details">
                                  <div class="v-messages theme--light">
                                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div class="col col-4">
                            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--switch primary--text">
                              <div class="v-input__control">
                                <div class="v-input__slot">
                                  <div class="v-input--selection-controls__input"><input aria-checked="true" id="input-139" role="switch" type="checkbox" aria-disabled="false" data-cy="account-on-budget" value="">
                                    <div class="v-input--selection-controls__ripple primary--text"></div>
                                    <div class="v-input--switch__track theme--light primary--text"></div>
                                    <div class="v-input--switch__thumb theme--light primary--text">
                                      <transition-stub name="fab-transition" mode="out-in"></transition-stub>
                                    </div>
                                  </div><label for="input-139" class="v-label theme--light" style="left: 0px; position: relative;">On Budget?</label>
                                </div>
                                <div class="v-messages theme--light primary--text">
                                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div class="col col-4">
                            <div class="v-input theme--light v-input--selection-controls v-input--switch">
                              <div class="v-input__control">
                                <div class="v-input__slot">
                                  <div class="v-input--selection-controls__input"><input aria-checked="false" id="input-143" role="switch" type="checkbox" aria-disabled="false" value="">
                                    <div class="v-input--selection-controls__ripple"></div>
                                    <div class="v-input--switch__track theme--light"></div>
                                    <div class="v-input--switch__thumb theme--light">
                                      <transition-stub name="fab-transition" mode="out-in"></transition-stub>
                                    </div>
                                  </div><label for="input-143" class="v-label theme--light" style="left: 0px; position: relative;">Invert Balance</label>
                                </div>
                                <div class="v-messages theme--light">
                                  <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                                </div>
                              </div>
                            </div> <span>(generally for credit card accounts)</span>
                          </div>
                          <div class="col col-12">
                            <div class="v-input v-textarea v-textarea--auto-grow v-textarea--no-resize v-input--dense theme--light v-text-field">
                              <div class="v-input__control">
                                <div class="v-input__slot">
                                  <div class="v-text-field__slot"><label for="input-147" class="v-label theme--light" style="left: 0px; position: absolute;">Notes</label><textarea data-cy="account-notes" id="input-147" rows="3"></textarea></div>
                                </div>
                                <div class="v-text-field__details">
                                  <div class="v-messages theme--light">
                                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                          <div class="col-sm-6 col-md-4 col-12">
                            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--prefix">
                              <div class="v-input__control">
                                <div class="v-input__slot">
                                  <div class="v-text-field__slot"><label for="input-150" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Starting Balance</label>
                                    <div class="v-text-field__prefix">$</div><input data-cy="account-starting-balance" id="input-150" type="text">
                                  </div>
                                </div>
                                <div class="v-text-field__details">
                                  <div class="v-messages theme--light">
                                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </form>
                    </div>
                    <hr role="separator" aria-orientation="horizontal" class="v-divider theme--light">
                    <div class="v-card__actions  white--text">
                      <div class="spacer"></div> <button type="button" class="v-btn v-btn--flat v-btn--text theme--light v-size--default" data-cy="account-cancel"><span class="v-btn__content">
        Cancel
      </span></button> <button type="button" class="v-btn v-btn--contained v-btn--is-elevated v-btn--has-bg theme--light v-size--default accent" id="saveAccountBtn"><span class="v-btn__content">
        Save
      </span></button>
                    </div>
                  </div>
                </form>
              </div>
            </transition-stub>
          </div>
        </div> <button type="button" class="mb-2 v-btn v-btn--contained v-btn--is-elevated v-btn--has-bg theme--dark v-size--default accent" id="addAccountBtn"><span class="v-btn__content">
          Add Account
        </span></button>
      </div>
    </header>
    <hr role="separator" aria-orientation="horizontal" class="pb-4 v-divider theme--light">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="Name" class="text-left"><span>Name</span></th>
            <th role="columnheader" scope="col" aria-label="Type: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Type</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="On Budget: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>On Budget</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Invert Balance: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Invert Balance</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Closed: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>Closed</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="Actions" class="text-start"><span>Actions</span></th>
          </tr>
        </thead>
        <tbody>
          <tr class="">
            <td class="text-left">Savings</td>
            <td class="text-start">CHECKING</td>
            <td class="text-start">true</td>
            <td class="text-start">false</td>
            <td class="text-start">false</td>
            <td class="text-start">
              <div class="crud-actions"><button type="button" icon="" class="v-icon notranslate v-icon--link material-icons theme--dark primary--text">edit</button> <button type="button" icon="" class="v-icon notranslate ml-1 v-icon--link material-icons theme--dark accent--text">delete</button></div>
            </td>
          </tr>
          <tr class="">
            <td class="text-left">Chase</td>
            <td class="text-start">CREDIT</td>
            <td class="text-start">true</td>
            <td class="text-start">true</td>
            <td class="text-start">false</td>
            <td class="text-start">
              <div class="crud-actions"><button type="button" icon="" class="v-icon notranslate v-icon--link material-icons theme--dark primary--text">edit</button> <button type="button" icon="" class="v-icon notranslate ml-1 v-icon--link material-icons theme--dark accent--text">delete</button></div>
            </td>
          </tr>
          <tr class="">
            <td class="text-left">Vanguard</td>
            <td class="text-start">INVESTMENT</td>
            <td class="text-start">false</td>
            <td class="text-start">false</td>
            <td class="text-start">false</td>
            <td class="text-start">
              <div class="crud-actions"><button type="button" icon="" class="v-icon notranslate v-icon--link material-icons theme--dark primary--text">edit</button> <button type="button" icon="" class="v-icon notranslate ml-1 v-icon--link material-icons theme--dark accent--text">delete</button></div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
